stl/inc/memory

```diff
@@ -2406,26 +2427,269 @@ bool atomic_compare_exchange_weak_explicit(shared_ptr<_Ty>* _Ptr, shared_ptr<_Ty
 }
 
 template <class _Ty>
-bool atomic_compare_exchange_weak(shared_ptr<_Ty>* _Ptr, shared_ptr<_Ty>* _Exp, shared_ptr<_Ty> _Value) {
+_CXX20_DEPRECATE_OLD_SHARED_PTR_ATOMIC_SUPPORT bool atomic_compare_exchange_weak(
+    shared_ptr<_Ty>* _Ptr, shared_ptr<_Ty>* _Exp, shared_ptr<_Ty> _Value) {
     // atomically compare and exchange
     return _STD atomic_compare_exchange_weak_explicit(
         _Ptr, _Exp, _STD move(_Value), memory_order_seq_cst, memory_order_seq_cst);
 }
 
 template <class _Ty>
-bool atomic_compare_exchange_strong_explicit(shared_ptr<_Ty>* _Ptr, shared_ptr<_Ty>* _Exp, shared_ptr<_Ty> _Value,
-    memory_order, memory_order) { // atomically compare and exchange
+_CXX20_DEPRECATE_OLD_SHARED_PTR_ATOMIC_SUPPORT bool atomic_compare_exchange_strong_explicit(shared_ptr<_Ty>* _Ptr,
+    shared_ptr<_Ty>* _Exp, shared_ptr<_Ty> _Value, memory_order, memory_order) { // atomically compare and exchange
     return _STD atomic_compare_exchange_weak_explicit(
         _Ptr, _Exp, _STD move(_Value), memory_order_seq_cst, memory_order_seq_cst);
 }
 
 template <class _Ty>
-bool atomic_compare_exchange_strong(shared_ptr<_Ty>* _Ptr, shared_ptr<_Ty>* _Exp, shared_ptr<_Ty> _Value) {
+_CXX20_DEPRECATE_OLD_SHARED_PTR_ATOMIC_SUPPORT bool atomic_compare_exchange_strong(
+    shared_ptr<_Ty>* _Ptr, shared_ptr<_Ty>* _Exp, shared_ptr<_Ty> _Value) {
     // atomically compare and exchange
     return _STD atomic_compare_exchange_strong_explicit(
         _Ptr, _Exp, _STD move(_Value), memory_order_seq_cst, memory_order_seq_cst);
 }
 
+#if _HAS_CXX20
+template <class _Ty>
+class alignas(2 * sizeof(void*)) _Atomic_ptr_base {
+    // overalignment is to allow potential future use of cmpxchg16b
+protected:
+    constexpr _Atomic_ptr_base() noexcept = default;
+
+    _Atomic_ptr_base(_Ty* const _Px, _Ref_count_base* const _Ref) noexcept
+        : _Ptr(_Px), _Repptr(reinterpret_cast<uintptr_t>(_Ref)) {}
+
+    _NODISCARD _Ref_count_base* _Lock_and_load() const noexcept {
+        constexpr uintptr_t _Low_bit = 1;
+        uintptr_t _Rep               = _Repptr.load() & ~_Low_bit;
+        while (!_Repptr.compare_exchange_weak(_Rep, _Rep | _Low_bit)) {
+            _Rep &= ~_Low_bit;
+            _YIELD_PROCESSOR();
+        }
+        return reinterpret_cast<_Ref_count_base*>(_Rep);
+    }
+
+    void _Store_and_unlock(_Ref_count_base* const _Value) const noexcept {
+        _Repptr.store(reinterpret_cast<uintptr_t>(_Value));

```

Comment:
Maybe add `std::memory_order_release` here and `std::memory_order_acquire` there?

Reply:
Consider `static_assert` on `alignas` to prove the use of low bit is fine.

Reply:
> Maybe add `std::memory_order_release` here and `std::memory_order_acquire` there?

Done

Reply:
> Consider `static_assert` on `alignas` to prove the use of low bit is fine.

I think test coverage is enough

Reply:
> Consider `static_assert` on `alignas` to prove the use of low bit is fine.

I think test coverage will prove that